File: backend/crdt/tree.go
package crdt

import "sort"

// TreeCRDT is an ordered-tree CRDT used for Slides (OFFICE-23). Each
// node has a stable id (OpID), a parent id, and an LWW "ordKey" used
// to position siblings. Move/reorder is an LWW write of the parent +
// ordKey. Deletion is a tombstone.
//
// Cycle prevention: a move that would create a cycle is dropped (the
// node keeps its previous parent). Cycles can only arise from
// concurrent moves; the loser is detected on Apply because we walk
// parent chains before accepting.
//
// Content of each slide is stored as a separate value string with its
// own LWW stamp. (For richer slide bodies the consumer can store a
// nested TextCRDT id; out of scope for this slice.)
type TreeCRDT struct {
	nodes map[OpID]*treeNode
	root  OpID
}

type treeNode struct {
	id      OpID
	parent  OpID
	ordKey  string // sortable position key, LWW
	ordID   OpID   // OpID of the last move/insert that set parent+ordKey
	value   string
	valueID OpID
	deleted bool
}

// NewTreeCRDT returns a tree with an implicit root (zero OpID).
func NewTreeCRDT() *TreeCRDT {
	return &TreeCRDT{nodes: map[OpID]*treeNode{}, root: OpID{}}
}

// TreeOpKind enumerates tree mutations.
type TreeOpKind uint8

const (
	TreeOpInsert  TreeOpKind = 1
	TreeOpMove    TreeOpKind = 2
	TreeOpSetText TreeOpKind = 3
	TreeOpDelete  TreeOpKind = 4
)

// TreeOp is a single tree mutation. OrdKey is a string so callers can
// use fractional-indexing schemes (e.g. "M", "MM", "MN") without this
// package prescribing one.
type TreeOp struct {
	Kind   TreeOpKind `json:"k"`
	ID     OpID       `json:"id"`
	Target OpID       `json:"t,omitempty"`
	Parent OpID       `json:"p,omitempty"`
	OrdKey string     `json:"o,omitempty"`
	Value  string     `json:"v,omitempty"`
}

// Apply applies op idempotently with LWW semantics on parent/ordKey
// and on value.
func (t *TreeCRDT) Apply(op TreeOp) {
	switch op.Kind {
	case TreeOpInsert:
		n, ok := t.nodes[op.ID]
		if ok {
			// A placeholder may have been created by an early-arriving
			// SetText/Move/Delete. Fill in parent/ordKey from this
			// Insert if the existing entry has no positioning op (or
			// has one that loses LWW to op.ID).
			if n.ordID.Equal(OpID{}) || op.ID.Less(n.ordID) == false {
				if !n.ordID.Equal(op.ID) {
					n.parent = op.Parent
					n.ordKey = op.OrdKey
					n.ordID = op.ID
				}
			}
			return
		}
		t.nodes[op.ID] = &treeNode{
			id:     op.ID,
			parent: op.Parent,
			ordKey: op.OrdKey,
			ordID:  op.ID,
		}
	case TreeOpMove:
		n, ok := t.nodes[op.Target]
		if !ok {
			// Buffer-via-tombstone so a late insert of Target merges
			// correctly. The insert won't overwrite because it bails
			// when id already exists; we'll just adopt parent/ordKey
			// from this move on insert as a small refinement:
			t.nodes[op.Target] = &treeNode{
				id: op.Target, parent: op.Parent, ordKey: op.OrdKey, ordID: op.ID,
			}
			return
		}
		// LWW: keep current if its ordID >= op.ID.
		if n.ordID.Equal(op.ID) || op.ID.Less(n.ordID) {
			return
		}
		if t.wouldCycle(op.Target, op.Parent) {
			return
		}
		n.parent = op.Parent
		n.ordKey = op.OrdKey
		n.ordID = op.ID
	case TreeOpSetText:
		n, ok := t.nodes[op.Target]
		if !ok {
			n = &treeNode{id: op.Target}
			t.nodes[op.Target] = n
		}
		if n.valueID.Equal(op.ID) || op.ID.Less(n.valueID) {
			return
		}
		n.value = op.Value
		n.valueID = op.ID
	case TreeOpDelete:
		n, ok := t.nodes[op.Target]
		if !ok {
			t.nodes[op.Target] = &treeNode{id: op.Target, deleted: true}
			return
		}
		n.deleted = true
	}
}

// wouldCycle reports whether re-parenting node under newParent would
// produce a cycle. The walk is bounded by the current node count to
// stay finite even if the tree is temporarily inconsistent (e.g. mid
// concurrent move arriving at the same node from two sides).
func (t *TreeCRDT) wouldCycle(node, newParent OpID) bool {
	cur := newParent
	limit := len(t.nodes) + 1
	for !cur.Equal(OpID{}) && limit > 0 {
		if cur.Equal(node) {
			return true
		}
		n, ok := t.nodes[cur]
		if !ok {
			return false
		}
		cur = n.parent
		limit--
	}
	// If we ran out of iterations we are in a transient cycle - treat
	// as "would cycle" so this move is dropped.
	return limit <= 0
}

// Children returns the non-deleted children of parent, sorted by
// (ordKey, OpID) for a stable order.
func (t *TreeCRDT) Children(parent OpID) []OpID {
	var out []OpID
	for id, n := range t.nodes {
		if !n.deleted && n.parent.Equal(parent) {
			out = append(out, id)
		}
	}
	sort.Slice(out, func(i, j int) bool {
		ni, nj := t.nodes[out[i]], t.nodes[out[j]]
		if ni.ordKey != nj.ordKey {
			return ni.ordKey < nj.ordKey
		}
		return out[i].Less(out[j])
	})
	return out
}

// Value returns the text value of node id (LWW).
func (t *TreeCRDT) Value(id OpID) (string, bool) {
	n, ok := t.nodes[id]
	if !ok || n.deleted {
		return "", false
	}
	return n.value, true
}

// Order returns a depth-first traversal of visible nodes starting at
// root, which for slides yields the rendered deck order.
func (t *TreeCRDT) Order() []OpID {
	var out []OpID
	t.walkOrder(t.root, &out)
	return out
}

func (t *TreeCRDT) walkOrder(parent OpID, out *[]OpID) {
	for _, c := range t.Children(parent) {
		*out = append(*out, c)
		t.walkOrder(c, out)
	}
}

type treeSnapshot struct {
	Nodes []treeNodeSnapshot `json:"nodes"`
}

type treeNodeSnapshot struct {
	ID      OpID   `json:"id"`
	Parent  OpID   `json:"p"`
	OrdKey  string `json:"o,omitempty"`
	OrdID   OpID   `json:"oi"`
	Value   string `json:"v,omitempty"`
	ValueID OpID   `json:"vi"`
	Deleted bool   `json:"d,omitempty"`
}

func (t *TreeCRDT) snapshot() treeSnapshot {
	out := treeSnapshot{Nodes: make([]treeNodeSnapshot, 0, len(t.nodes))}
	for _, n := range t.nodes {
		out.Nodes = append(out.Nodes, treeNodeSnapshot{
			ID: n.id, Parent: n.parent, OrdKey: n.ordKey, OrdID: n.ordID,
			Value: n.value, ValueID: n.valueID, Deleted: n.deleted,
		})
	}
	sort.Slice(out.Nodes, func(i, j int) bool {
		return out.Nodes[i].ID.Less(out.Nodes[j].ID)
	})
	return out
}

func (t *TreeCRDT) restore(s treeSnapshot) {
	t.nodes = map[OpID]*treeNode{}
	for _, n := range s.Nodes {
		t.nodes[n.ID] = &treeNode{
			id: n.ID, parent: n.Parent, ordKey: n.OrdKey, ordID: n.OrdID,
			value: n.Value, valueID: n.ValueID, deleted: n.Deleted,
		}
	}
}

File: src/lib/crdt/tree.js
/**
 * src/lib/crdt/tree.js
 *
 * Browser-side LWW ordered-tree CRDT for Slides (OFFICE-23).
 *
 * Mirrors backend/crdt/tree.go:
 *   - TreeOpInsert  — add a new slide node under root
 *   - TreeOpMove    — reorder / reparent (LWW on ordKey + parent)
 *   - TreeOpSetText — update slide content (LWW on value)
 *   - TreeOpDelete  — tombstone a slide
 *
 * Slide content is stored as a JSON-encoded value string per node
 * (matching how SlidesEditor represents each slide as an object).
 *
 * Usage
 * -----
 *   import { TreeSession } from './crdt/tree.js';
 *
 *   const session = new TreeSession({ sessionId: fileId, replicaId, fabricClient });
 *   session.addEventListener('remoteOp', () => { ... rerender ... });
 *
 *   // Local ops:
 *   const nodeId = session.insertSlide(ordKey, slideData);
 *   session.setSlide(nodeId, slideData);
 *   session.moveSlide(nodeId, newOrdKey);
 *   session.deleteSlide(nodeId);
 *
 *   // Read ordered slides:
 *   const slides = session.orderedSlides(); // → [{ nodeId, data }]
 *
 *   session.destroy();
 */

// ---------------------------------------------------------------------------
// Lamport clock (same as grid.js)
// ---------------------------------------------------------------------------

class LamportClock {
  constructor(replicaId) {
    this.replicaId = replicaId
    this.c = 0
  }

  tick() {
    this.c += 1
    return this._format(this.c)
  }

  observe(remoteCounter) {
    if (remoteCounter > this.c) this.c = remoteCounter
  }

  _format(counter) {
    return (
      String(Date.now()).padStart(20, '0') +
      '_' +
      String(counter).padStart(10, '0') +
      '_' +
      this.replicaId
    )
  }
}

function opIdLess(a, b) {
  const [, ac, ar] = a.split('_')
  const [, bc, br] = b.split('_')
  const ai = parseInt(ac, 10)
  const bi = parseInt(bc, 10)
  if (ai !== bi) return ai < bi
  return ar < br
}

// ---------------------------------------------------------------------------
// TreeOp kinds (matches backend/crdt/tree.go)
// ---------------------------------------------------------------------------

const TREE_OP_INSERT   = 1
const TREE_OP_MOVE     = 2
const TREE_OP_SET_TEXT = 3
const TREE_OP_DELETE   = 4

// The implicit root has this nodeId.
const ROOT_ID = ''

// ---------------------------------------------------------------------------
// TreeCRDT — in-memory LWW ordered tree
// ---------------------------------------------------------------------------

class TreeCRDT {
  constructor() {
    // nodeId (string) → { id, parent, ordKey, ordId, value, valueId, deleted }
    this._nodes = new Map()
  }

  apply(op) {
    switch (op.kind) {
      case TREE_OP_INSERT: {
        if (this._nodes.has(op.id)) {
          // Node exists; fill parent/ordKey if we don't have a positioning op yet.
          const n = this._nodes.get(op.id)
          if (!n.ordId || opIdLess(n.ordId, op.id)) {
            n.parent = op.parent
            n.ordKey = op.ordKey
            n.ordId  = op.id
          }
          return
        }
        this._nodes.set(op.id, {
          id: op.id, parent: op.parent, ordKey: op.ordKey, ordId: op.id,
          value: '', valueId: '', deleted: false,
        })
        break
      }
      case TREE_OP_MOVE: {
        let n = this._nodes.get(op.target)
        if (!n) {
          // Buffer node for late-arriving Insert.
          this._nodes.set(op.target, {
            id: op.target, parent: op.parent, ordKey: op.ordKey, ordId: op.id,
            value: '', valueId: '', deleted: false,
          })
          return
        }
        // LWW: keep current if its ordId >= op.id.
        if (n.ordId && !opIdLess(op.id, n.ordId) && op.id !== n.ordId) {
          if (this._wouldCycle(op.target, op.parent)) return
          n.parent = op.parent
          n.ordKey = op.ordKey
          n.ordId  = op.id
        } else if (!n.ordId) {
          if (this._wouldCycle(op.target, op.parent)) return
          n.parent = op.parent
          n.ordKey = op.ordKey
          n.ordId  = op.id
        }
        break
      }
      case TREE_OP_SET_TEXT: {
        let n = this._nodes.get(op.target)
        if (!n) {
          n = { id: op.target, parent: ROOT_ID, ordKey: '', ordId: '', value: '', valueId: '', deleted: false }
          this._nodes.set(op.target, n)
        }
        if (!n.valueId || opIdLess(n.valueId, op.id)) {
          n.value   = op.value
          n.valueId = op.id
        }
        break
      }
      case TREE_OP_DELETE: {
        let n = this._nodes.get(op.target)
        if (!n) {
          this._nodes.set(op.target, {
            id: op.target, parent: ROOT_ID, ordKey: '', ordId: '', value: '', valueId: '', deleted: true,
          })
          return
        }
        n.deleted = true
        break
      }
    }
  }

  _wouldCycle(node, newParent) {
    let cur = newParent
    let limit = this._nodes.size + 1
    while (cur && cur !== ROOT_ID && limit-- > 0) {
      if (cur === node) return true
      const n = this._nodes.get(cur)
      if (!n) return false
      cur = n.parent
    }
    return limit <= 0
  }

  /** Return visible children of parent sorted by (ordKey, id). */
  children(parentId) {
    const out = []
    for (const [id, n] of this._nodes) {
      if (!n.deleted && n.parent === parentId) out.push(id)
    }
    out.sort((a, b) => {
      const na = this._nodes.get(a)
      const nb = this._nodes.get(b)
      if (na.ordKey !== nb.ordKey) return na.ordKey < nb.ordKey ? -1 : 1
      return a < b ? -1 : a > b ? 1 : 0
    })
    return out
  }

  /** Depth-first ordered list of visible node ids from root. */
  order() {
    const out = []
    this._walk(ROOT_ID, out)
    return out
  }

  _walk(parentId, out) {
    for (const id of this.children(parentId)) {
      out.push(id)
      this._walk(id, out)
    }
  }

  value(id) {
    const n = this._nodes.get(id)
    if (!n || n.deleted) return undefined
    return n.value
  }

  snapshot() {
    const out = []
    for (const [, n] of this._nodes) {
      out.push({ id: n.id, parent: n.parent, ordKey: n.ordKey, ordId: n.ordId,
                 value: n.value, valueId: n.valueId, deleted: n.deleted })
    }
    return out
  }

  restore(nodes) {
    this._nodes.clear()
    for (const n of nodes) {
      this._nodes.set(n.id, { ...n })
    }
  }
}

// ---------------------------------------------------------------------------
// Fractional-index helper (simple string-based)
// ---------------------------------------------------------------------------

/**
 * Return an ordKey string positioned between `before` and `after`.
 * Uses a simple midpoint-string approach; infinite precision.
 */
export function ordKeyBetween(before, after) {
  const b = before || 'a'
  const a = after  || 'z'
  if (b < a) {
    // Return a string midway between the two.
    const mid = midString(b, a)
    if (mid && mid > b && mid < a) return mid
  }
  // Fallback: append 'm' to before.
  return b + 'm'
}

function midString(lo, hi) {
  // Find first differing position.
  let i = 0
  while (i < lo.length && i < hi.length && lo[i] === hi[i]) i++
  if (i >= lo.length) {
    // lo is a prefix of hi — insert lo + mid char.
    const hc = hi.charCodeAt(i)
    const ac = 'a'.charCodeAt(0)
    if (hc > ac + 1) return lo + String.fromCharCode(Math.floor((ac + hc) / 2))
    return lo + 'a' + midChar(hi[i + 1])
  }
  const lc = lo.charCodeAt(i) || 'a'.charCodeAt(0)
  const hc = hi.charCodeAt(i)
  if (hc - lc > 1) return lo.slice(0, i) + String.fromCharCode(Math.floor((lc + hc) / 2))
  return lo + 'm'
}

function midChar(c) {
  const code = c ? c.charCodeAt(0) : 'z'.charCodeAt(0)
  return String.fromCharCode(Math.floor(('a'.charCodeAt(0) + code) / 2))
}

// ---------------------------------------------------------------------------
// TreeSession — ties TreeCRDT to a FabricClient
// ---------------------------------------------------------------------------

const SNAPSHOT_KEY = (id) => `crdt_tree_${id}`
const OP_LOG_KEY   = (id) => `crdt_tree_ops_${id}`
const MAX_OPLOG    = 500

export class TreeSession extends EventTarget {
  /**
   * @param {object} opts
   * @param {string}            opts.sessionId
   * @param {string}            opts.replicaId
   * @param {FabricClient|null} [opts.fabricClient]
   */
  constructor({ sessionId, replicaId, fabricClient = null }) {
    super()
    this._session   = sessionId
    this._replicaId = replicaId
    this._fabric    = fabricClient
    this._clock     = new LamportClock(replicaId)
    this._crdt      = new TreeCRDT()
    this._destroyed = false

    this._loadLocal()

    if (this._fabric) {
      this._onFabricMessage = (ev) => this._handleFabricMessage(ev.detail.data)
      this._fabric.addEventListener('message', this._onFabricMessage)
    }
  }

  // -------------------------------------------------------------------------
  // Local mutations
  // -------------------------------------------------------------------------

  /**
   * Insert a new slide node.
   * @param {string}  ordKey   - position key (use ordKeyBetween)
   * @param {object}  data     - slide data object (will be JSON-encoded as value)
   * @returns {string} the new nodeId
   */
  insertSlide(ordKey, data) {
    const id = this._clock.tick()
    const op = { kind: TREE_OP_INSERT, id, parent: ROOT_ID, ordKey }
    this._crdt.apply(op)
    this._broadcast({ type: 'tree_op', session: this._session, op })
    this._persistOp(op)

    // Set initial content.
    if (data !== undefined) {
      const setOp = { kind: TREE_OP_SET_TEXT, id: this._clock.tick(), target: id, value: JSON.stringify(data) }
      this._crdt.apply(setOp)
      this._broadcast({ type: 'tree_op', session: this._session, op: setOp })
      this._persistOp(setOp)
    }
    return id
  }

  /** Update the content of an existing slide node. */
  setSlide(nodeId, data) {
    const id = this._clock.tick()
    const op = { kind: TREE_OP_SET_TEXT, id, target: nodeId, value: JSON.stringify(data) }
    this._crdt.apply(op)
    this._broadcast({ type: 'tree_op', session: this._session, op })
    this._persistOp(op)
  }

  /** Move / reorder a slide. */
  moveSlide(nodeId, newOrdKey) {
    const id = this._clock.tick()
    const op = { kind: TREE_OP_MOVE, id, target: nodeId, parent: ROOT_ID, ordKey: newOrdKey }
    this._crdt.apply(op)
    this._broadcast({ type: 'tree_op', session: this._session, op })
    this._persistOp(op)
  }

  /** Delete a slide. */
  deleteSlide(nodeId) {
    const id = this._clock.tick()
    const op = { kind: TREE_OP_DELETE, id, target: nodeId }
    this._crdt.apply(op)
    this._broadcast({ type: 'tree_op', session: this._session, op })
    this._persistOp(op)
  }

  // -------------------------------------------------------------------------
  // Query
  // -------------------------------------------------------------------------

  /**
   * Return ordered slides as [{ nodeId, data }].
   * data is the parsed slide object stored by insertSlide / setSlide.
   */
  orderedSlides() {
    return this._crdt.order().map((nodeId) => {
      const raw = this._crdt.value(nodeId)
      let data
      try { data = raw ? JSON.parse(raw) : {} } catch { data = {} }
      return { nodeId, data }
    })
  }

  // -------------------------------------------------------------------------
  // Persistence (localStorage)
  // -------------------------------------------------------------------------

  saveLocal() {
    try {
      localStorage.setItem(SNAPSHOT_KEY(this._session), JSON.stringify(this._crdt.snapshot()))
    } catch { /* quota — ignore */ }
  }

  _loadLocal() {
    try {
      const raw = localStorage.getItem(SNAPSHOT_KEY(this._session))
      if (raw) {
        const nodes = JSON.parse(raw)
        this._crdt.restore(nodes)
        for (const n of nodes) {
          for (const opId of [n.ordId, n.valueId]) {
            if (opId) {
              const parts = opId.split('_')
              this._clock.observe(parseInt(parts[1], 10) || 0)
            }
          }
        }
      }
      const logRaw = localStorage.getItem(OP_LOG_KEY(this._session))
      if (logRaw) {
        const ops = JSON.parse(logRaw)
        for (const op of ops) this._crdt.apply(op)
      }
    } catch { /* corrupt — ignore */ }
  }

  _persistOp(op) {
    try {
      const logRaw = localStorage.getItem(OP_LOG_KEY(this._session))
      const ops = logRaw ? JSON.parse(logRaw) : []
      ops.push(op)
      if (ops.length > MAX_OPLOG) ops.splice(0, ops.length - MAX_OPLOG)
      localStorage.setItem(OP_LOG_KEY(this._session), JSON.stringify(ops))
    } catch { /* quota — ignore */ }
  }

  // -------------------------------------------------------------------------
  // Fabric transport
  // -------------------------------------------------------------------------

  _broadcast(msg) {
    if (!this._fabric) return
    try {
      this._fabric.send(JSON.stringify(msg))
    } catch { /* disconnected */ }
  }

  _handleFabricMessage(raw) {
    if (this._destroyed) return
    let msg
    try { msg = typeof raw === 'string' ? JSON.parse(raw) : raw } catch { return }
    if (!msg || msg.session !== this._session) return

    if (msg.type === 'tree_op' && msg.op) {
      const op = msg.op
      // Advance clock.
      for (const field of [op.id, op.target]) {
        if (field && typeof field === 'string') {
          const parts = field.split('_')
          this._clock.observe(parseInt(parts[1], 10) || 0)
        }
      }
      this._crdt.apply(op)
      this._persistOp(op)
      this.dispatchEvent(new CustomEvent('remoteOp', { detail: { op } }))
    } else if (msg.type === 'tree_snapshot_request') {
      this._broadcast({
        type: 'tree_snapshot',
        session: this._session,
        nodes: this._crdt.snapshot(),
      })
    } else if (msg.type === 'tree_snapshot' && msg.nodes) {
      for (const n of msg.nodes) {
        if (n.ordId) {
          this._crdt.apply({ kind: TREE_OP_INSERT, id: n.ordId, parent: n.parent, ordKey: n.ordKey })
        }
        if (n.valueId && n.value) {
          this._crdt.apply({ kind: TREE_OP_SET_TEXT, id: n.valueId, target: n.id, value: n.value })
        }
        if (n.deleted) {
          this._crdt.apply({ kind: TREE_OP_DELETE, id: n.ordId || n.id, target: n.id })
        }
      }
      this.saveLocal()
      this.dispatchEvent(new CustomEvent('remoteOp', { detail: { snapshot: true } }))
    }
  }

  /** Request a snapshot from peers on first join. */
  requestSnapshot() {
    this._broadcast({ type: 'tree_snapshot_request', session: this._session })
  }

  destroy() {
    this._destroyed = true
    if (this._fabric && this._onFabricMessage) {
      this._fabric.removeEventListener('message', this._onFabricMessage)
    }
    this.saveLocal()
  }
}

// ---------------------------------------------------------------------------
// Stable per-tab replicaId
// ---------------------------------------------------------------------------

export function getTreeReplicaId() {
  let id = sessionStorage.getItem('crdt_tree_replica')
  if (!id) {
    id = crypto.randomUUID().slice(0, 8)
    sessionStorage.setItem('crdt_tree_replica', id)
  }
  return id
}
